src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,120 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] array;
+    private int size;
+    private int minCapacity;
+
+    public ArrayList() {
+        this.array = (T[]) new Object[DEFAULT_CAPACITY];
+        this.size = 0;
+        this.minCapacity = DEFAULT_CAPACITY;
+    }
+
     @Override
     public void add(T value) {
-
+        this.array = grow(minCapacity);
+        array[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexForAddMethod(index);
+        if (index + 1 <= minCapacity) {
+            this.array = grow(minCapacity);
+            T[] newArray = grow(minCapacity);
+            System.arraycopy(this.array, 0, newArray, 0, index);
+            newArray[index] = value;
+            System.arraycopy(this.array, index, newArray, index + 1, minCapacity - index - 1);
+            copyArray(newArray, this.array);
+            size++;
+        } else {
+            add(value);
+        }
+    }
 
+    private T[] grow(int minCapacity) {
+        T[] arrayWithNewCapacity = (T[]) new Object[getNewCapacity(minCapacity)];
+        copyArray(this.array, arrayWithNewCapacity);
+        return arrayWithNewCapacity;
     }
 
-    @Override
-    public void addAll(List<T> list) {
+    private int getNewCapacity(int minCapacity) {
+        if (size + 1 > array.length) {
+            return this.minCapacity = minCapacity + (minCapacity >> 1);
+        }
+        return minCapacity;
+    }
 
+    private void checkIndexForAddMethod(int index) {
```

Comment:
As you use this method only ones we can move this logic to `add` method

Reply:
removed